### PR TITLE
JarInfer: Third-party bytecode analysis (MVP version)

### PR DESCRIPTION
**Pull request by lazaroclapp**
Initial merge of JarInfer third-party bytecode analysis support for NullAway.

- Adds jar-infer-cli tool, which uses WALA to analyze public/protected methods of bytecode files.
- The jar-infer-cli tool produces an output Jar, which is the original analyzed Jar with added metadata.
- The analysis' metadata marks methods as having `@Nullable` return if they include any `return null;` statement and arguments as `@NonNull` if they are dereferenced in a basic block that post-dominates the method’s entry block.
- We add a NullAway handler that looks at this metadata and uses it to override NullAway's default (optimistic) assumptions about third-party code.

### CODE REVIEW

File: jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/DefinitelyDerefedParams.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+import com.ibm.wala.ssa.SSAGetInstruction;
+import com.ibm.wala.ssa.SSAInstruction;
+import com.ibm.wala.ssa.SSAPutInstruction;
+import com.ibm.wala.ssa.SSAReturnInstruction;
+import com.ibm.wala.util.collections.Iterator2Iterable;
+import com.ibm.wala.util.graph.Graph;
+import com.ibm.wala.util.graph.GraphUtil;
+import com.ibm.wala.util.graph.dominators.Dominators;
+import com.ibm.wala.util.graph.impl.GraphInverter;
+import java.util.*;
```
Comment (msridhar) [inline jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/DefinitelyDerefedParams.java]: Can we get rid of the `*` import?

File: settings.gradle [hunk 2]
```diff
@@ -3,4 +3,8 @@ include ':sample-library-model'
 include ':sample'
 include ':sample-app'
 include ':test-java-lib'
-include ':compile-bench'
\ No newline at end of file
+include ':test-java-lib-jarinfer'
```
Comment (msridhar) [inline settings.gradle]: Maybe not for this PR, but it'd be nice to not have all the modules at the top-level and to have the jarinfer modules in their own folder

File: jar-infer-cli/build.gradle [hunk 3]
```diff
@@ -0,0 +1,88 @@
+plugins {
+    id "net.ltgt.errorprone" version "0.0.13"
+    id "com.github.johnrengelman.shadow" version "2.0.2"
```
[the remark this answers is not in the document]
Comment (subarnob) [reply jar-infer-cli/build.gradle]: We're creating fat jar for the `jar-infer-cli` tool.

File: settings.gradle (hunk 2 above)
Comment (lazaroclapp) [reply settings.gradle]: @subarnob Can you take a look at how hard it would be to do this for this PR. If it's simple enough, I would love to have all these targets under a `jarinfer/` top level dir on the repo from day one :) (if it's hard to do, I can take a look later)

File: nullaway/src/main/java/com/uber/nullaway/handlers/CompositeHandler.java [hunk 4]
```diff
@@ -25,13 +25,7 @@
...
 import com.google.common.collect.ImmutableSet;
 import com.google.errorprone.VisitorState;
-import com.sun.source.tree.ClassTree;
-import com.sun.source.tree.ExpressionTree;
-import com.sun.source.tree.LambdaExpressionTree;
-import com.sun.source.tree.MemberReferenceTree;
-import com.sun.source.tree.MethodInvocationTree;
-import com.sun.source.tree.MethodTree;
-import com.sun.source.tree.ReturnTree;
+import com.sun.source.tree.*;
```
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/handlers/CompositeHandler.java]: Nit: star import

File: gradle/wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -1,3 +1,4 @@
+#Fri Jun 29 14:35:16 PDT 2018
```
Comment (msridhar) [inline gradle/wrapper/gradle-wrapper.properties]: Nit: unnecessary change

**Comment by lazaroclapp** [comment]
@msridhar @subarnob Went ahead and fixed the nits on top of this branch, to avoid another PR -> merge-to-jar-infer -> merge-to-master cycle :)
